kimchi/src/snarky/cvar.rs

```diff
@@ -0,0 +1,329 @@
+        let mut res = CVar::Constant(F::zero());
+        for (cst, term) in terms {
+            res = res.add(&term.scale(*cst));
+        }
+        res
+    }
+
+    pub fn sum(vs: &[&Self]) -> Self {
+        let terms: Vec<_> = vs.iter().map(|v| (F::one(), (*v).clone())).collect();
+        Self::linear_combination(&terms)
+    }
+
+    pub fn mul(&self, other: &Self, label: Option<&'static str>, cs: &mut RunState<F>) -> Self {
+        match (self, other) {
+            (CVar::Constant(x), CVar::Constant(y)) => CVar::Constant(*x * y),
+
+            // TODO: this was not in the original ocaml code, but seems correct to me
+            (CVar::Constant(cst), _) | (_, CVar::Constant(cst)) if cst.is_zero() => {
+                CVar::Constant(F::zero())
+            }
+
+            // TODO: same here
+            (CVar::Constant(cst), cvar) | (cvar, CVar::Constant(cst)) if cst.is_one() => {
+                cvar.clone()
+            }
+
+            (CVar::Constant(cst), cvar) | (cvar, CVar::Constant(cst)) => cvar.scale(*cst),
+
+            (_, _) => {
+                let self_clone = self.clone();
+                let other_clone = other.clone();
+                let res: CVar<F> = cs.compute(loc!(), move |env| {
+                    let x: F = env.read_var(&self_clone);
+                    let y: F = env.read_var(&other_clone);
+                    x * y
+                });
+
+                let label = label.or(Some("checked_mul"));
+
+                cs.assert_r1cs(label, self.clone(), other.clone(), res.clone());
+                res
+            }
+        }
+    }
+
+    /** [equal_constraints z z_inv r] asserts that
+       if z = 0 then r = 1, or
+       if z <> 0 then r = 0 and z * z_inv = 1
+    */
+    fn equal_constraints(state: &mut RunState<F>, z: Self, z_inv: Self, r: Self) {
+        // TODO: the ocaml code actually calls assert_all
+        let one_minus_r = CVar::Constant(F::one()) - &r;
+        let zero = CVar::Constant(F::zero());
+        state.assert_r1cs(Some("equals_1"), z_inv, z.clone(), one_minus_r);
+        state.assert_r1cs(Some("equals_2"), r, z, zero);
+    }
+
+    /** [equal_vars z] computes [(r, z_inv)] that satisfy the constraints in
+    [equal_constraints z z_inv r].
+
+    In particular, [r] is [1] if [z = 0] and [0] otherwise.
+    */
+    fn equal_vars(env: &dyn WitnessGeneration<F>, z: &CVar<F>) -> (F, F) {
+        let z: F = env.read_var(z);
+        if let Some(z_inv) = z.inverse() {
+            (F::zero(), z_inv)
+        } else {
+            (F::one(), F::zero())
+        }
+    }
+
+    pub fn equal(&self, state: &mut RunState<F>, other: &CVar<F>) -> Boolean<F> {
+        match (self, other) {
+            (CVar::Constant(x), CVar::Constant(y)) => {
+                let res = if x == y { F::one() } else { F::zero() };
+                let cvars = vec![CVar::Constant(res)];
+                Boolean::from_cvars_unsafe(cvars, ())
```

Comment:
As in the previous review, please create a `create_unsafe` or similar function for `Boolean` instead of using the internal one.